cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/token/SingleClusterReplicaAwareFailureHandler.java

```diff
@@ -0,0 +1,191 @@
+
+        if (isEmpty)
+        {
+            return failedRanges;
+        }
+
+        for (Map.Entry<Range<BigInteger>, FailuresPerInstance> failedRangeEntry : rangeFailuresMap.asMapOfRanges().entrySet())
+        {
+            Range<BigInteger> range = failedRangeEntry.getKey();
+            FailuresPerInstance errorMap = failedRangeEntry.getValue();
+            Set<I> failedReplicas = errorMap.instances()
+                                            .stream()
+                                            .filter(errorMap::hasError)
+                                            .collect(Collectors.toSet());
+
+            // no failures found for the range; skip consistency check on this one and move on
+            if (failedReplicas.isEmpty())
+            {
+                continue;
+            }
+
+            tokenRangeMapping.getWriteReplicasOfRange(range, instance -> {
+                                 boolean shouldKeep = localDC == null || instance.datacenter().equalsIgnoreCase(localDC);
+                                 if (shouldKeep && clusterId != null)
+                                 {
+                                     shouldKeep = clusterId.equalsIgnoreCase(instance.clusterId());
+                                 }
+                                 return shouldKeep;
+                             })
+                             .forEach((subrange, liveAndDown) -> {
+                                 if (!checkSubrange(cl, localDC, replicationFactor, liveAndDown, failedReplicas))
+                                 {
+                                     failedRanges.add(new ConsistencyFailurePerRange(subrange, errorMap));
+                                 }
+                             });
+        }
+
+        return failedRanges;
+    }
+
+    /**
+     * Check whether a CL can be satisfied for each sub-range.
+     * @return true if consistency is satisfied; false otherwise.
+     */
+    private boolean checkSubrange(ConsistencyLevel cl,
+                                  @Nullable String localDC,
+                                  ReplicationFactor replicationFactor,
+                                  Set<I> liveAndDown,
+                                  Set<I> failedReplicas)
+    {
+        Set<I> liveReplicas = liveAndDown.stream()
+                                         .filter(instance -> instance.nodeStatus() == NodeStatus.UP)
+                                         .collect(Collectors.toSet());
+        Set<I> pendingReplicas = liveAndDown.stream()
+                                            .filter(instance -> instance.nodeState().isPending)
+                                            .collect(Collectors.toSet());
+        // success is assumed if not failed
+        Set<I> succeededReplicas = liveReplicas.stream()
+                                               .filter(instance -> !failedReplicas.contains(instance))
+                                               .collect(Collectors.toSet());
```

Comment:
This seems unnecessarily complex, maybe simplify?
```suggestion
                Set<I> pendingReplicas = new HashSet<>();
        // success is assumed if not failed
        Set<I> succeededReplicas = new HashSet<>();
        for (I instance : liveAndDown)
        {
            if (instance.nodeStatus() == NodeStatus.UP && !failedReplicas.contains(instance))
            {
                succeededReplicas.add(instance);
            }
            if (instance.nodeState().isPending)
            {
                pendingReplicas.add(instance);
            }
        }
```

Reply:
ok. it is the same code as before. I can update to using for-loop.